enhancements/authentication/improved-login-workflow.md

```diff
@@ -0,0 +1,245 @@
+be done as prescribed by the RFC.
+
+#### OAuth server
+
+The internal OAuth server uses [OSIN](https://github.com/openshift/osin) as a dependency to provide OAuth2 Authorized
+Code Grant Flow for a registered client to fetch tokens. It also has support for PKCE(Proof Key for Code Exchange) as
+specified in [RFC-7636](https://datatracker.ietf.org/doc/html/rfc7636). _RFC-7636_ is an enhancement for public clients,
+i.e. clients which cannot hold client credentials to fetch an access token securely. Similarly `oc` is also a public
+client because it cannot be distributed with client credentials to retrieve the access token and PKCE is designed for
+use cases like this one.
+
+The OSIN library will have to be modified so that when the redirect URI is a loopback address of the
+form `http://127.0.0.1:{port}/{path}` or `http://[::1]:{port}/{path}` the redirect URL verification should ignore the
+port value. This is required because one particular port may not always be available on all machines where _oc_ is run.
+
+#### cluster-authentication-operator
+
+The [cluster-authentication-operator](https://github.com/openshift/cluster-authentication-operator) provisions the
+default [OAuthClient](https://docs.openshift.com/container-platform/4.9/rest_api/oauth_apis/oauthclient-oauth-openshift-io-v1.html)
+resources. It already provisions a client for `oc` but sets the `respondWithChallenges` to _true_ which is unsuitable
+for OAuth Grant Flow. Therefore, another client is required where the `respondWithChallenges` is set to `false` which
+should also be provisioned by the operator.
+
+### Risks and Mitigations
+
+1. There are no changes to existing behavior for users of `oc`. Previous login workflows continue working with the same
+   behavior as in previous versions.
+2. The new login workflow in _oc_ will need to listen on a port for the callback from the OAuth server. This port cannot
+   be fixed because the port may be in-use on the user's machine and the user may not have the permissions to resolve
+   this. Instead, the user should be able to specify an alternate port which is free on the system. If the user does not
+   specify a port _oc_ can pick any available port at random.
+4. Some OSes under certain configurations might prevent the callback server from listening on the loopback interface. In
+   this case the server would not start and `oc` would print a message that it cannot listen on that interface.
+5. If `127.0.0.1` on the user's machine is not the machine itself but has been compromised and points to another
+   location then the attacker can obtain the security code. But this code cannot be exchanged for a token because the
+   attacker does not have access to the challenge key which is only in memory on the users machine.
+
+## Design Details
+
+### Open Questions
+
+1. ~~Should we modify the OAuth server library that it allows any port for the redirect URL when the url is localhost?~~
+   Yes to be in line with RFC-8252.
```

Comment:
it's not an open question, then :slightly_smiling_face:

Reply:
Actually the double tildes at the beginning is a strike-through. It's not visible in preview. I will remove them but I wanted to keep them in some form in the enhancement.

Reply:
Ah, true, I ignored the tildes :slightly_smiling_face: